### DISCUSSION TITLE
Rust crates & more permissive licensing

### DISCUSSION THREAD

**Discussion by larrysalibra** [stacks-network/stacks-core#3659] [General]
I was looking for Stacks-related libraries to use in a Rust project...similar functionality to what exists in stacks.js. There aren't any crates (that I was able to find) with code relevant to Stacks. In comparison there are comparatively large ecosystems of bitcoin, ethereum, lightning-related crates available.

A good deal of code in this repo such as the code paths used by the CLI - provides this functionality. 

One could simply copy and paste that code into a useful Stacks rust library, but the fact that this repo is under GPLv3 makes that not very useful in that the resulting library would also have to be released under GPLv3, which would make it unusable in most projects.

There was a previous discussion about more permissive licensing here #223.

I think it might be worth revisiting.

This issue is to suggest either (or both) of the following:

1) refactoring code that has functionality similar to stack.js into a new library under a more permissive license that can be both consumed by this project and other applications and libraries

2) relicense this repo under a more permissive license so that the community can use existing code from this repo in libraries and/or apps of their own creation.

**Comment by igorsyl** [comment]
stacks.rs should ship in a few weeks: https://github.com/stacksgov/Stacks-Grant-Launchpad/issues/869

@will-at-stacks: @larrysalibra has a good point on licensing. We should review all licensing requirements for the grants. Ideally MIT license?

**Comment by jcnelson** [comment]
It will remain GPLv3. This isn't a charity. If you want an exception for
your project, please contact us via email and we'll work out a licensing
fee.

**Comment by igorsyl** [comment]
Posting the following as I was confused about the difference between non-profit and charity:

"© 2023 Stacks Open Internet Foundation. Delaware Nonprofit Foundation."

"A nonprofit is an organization that uses its income and profits for the organization's main goal that supports the mission. On the other hand, a charity is a type of nonprofit that engages in activities aimed at improving lives in the communities."

**Comment by jcnelson** [comment]
> stacks.rs should ship in a few weeks: stacksgov/Stacks-Grant-Launchpad#869

@igorsyl If `stacks.rs` uses or statically links any code from this repo, then the `stacks.rs` components which do so must be licensed as GPLv3.  I hope the grant review committee knows this.

> One could simply copy and paste that code into a useful Stacks rust library, but the fact that this repo is under GPLv3 makes that not very useful in that the resulting library would also have to be released under GPLv3, which would make it unusable in most projects.

Such as?  "Most projects" usually means "my projects" in these conversations.  Can the other "most projects" maintainers please comment on how and why the use of GPLv3 prevents them from making useful products derived from code in this repository?

**Comment by pavitthrap** [comment]
Blockchain meeting: no action.

**Comment by kantai** [comment]
I'm going to convert this to a discussion, because this isn't directly addressable.

**Comment by kantai** [comment]
I think that there's probably something worth thinking about regarding crate licensing versus binary licensing.

LGPL would allow _linking_ the crate -- this would be useful for applications that want to use, e.g., the message parsing codebase.

**Comment by larrysalibra** [comment]
> We should review all licensing requirements for the grants. Ideally MIT license?

Most libraries in the Bitcoin ecosystem are permissively licensed - this repo itself benefits from that.

----

Not entirely sure what stacks.rs will include, but on the surface it sounds like what I'm personally looking for! Thanks for pointing me to https://github.com/stacksgov/Stacks-Grant-Launchpad/issues/869 @igorsyl 

---

There's also a mostly undocumented crate `stacks-common` here: https://crates.io/crates/stacks-common

Unclear if this is intended to be used by 3rd parties.

---

From https://github.com/Trust-Machines/core-eng/issues/64#issuecomment-1431445558

> Wonderful that we're looking into building stacks.rs. Some initial thoughts before we start to dig into the design and requirements:

> 1. We must be careful with licenses. Since we're aiming for widespread adoption we should have a permissive license, e.g. Apache2 > or  MIT. This means we can't depend on the reference implementation for Stacks which is GPL-licensed.

It seems that they are aware of the licensing issues w.r.t. stacks.rs. Because of the licensing of this repo, they need to write stacks.rs without using anything from stacks. This seems kind of silly to me. Unclear what problem that solves.
---

To be clear, I was working on a BNS library which I intended to be MIT licensed because I want any and all applications to use it. I'm not in the business of forcing licensing decisions on application developers...which is what GPLv3 licensed libraries do - all that succeeds in doing is reducing total addressable market.
 
---

> I think that there's probably something worth thinking about regarding crate licensing versus binary licensing.
> LGPL would allow linking the crate -- this would be useful for applications that want to use, e.g., the message parsing codebase. 

This makes sense to me!
